### PR TITLE
[1.14.x] Extendible categories in recipe book

### PR DESCRIPTION
**Pull request by yanny7**
Basicaly allows modders to add recipe book into custom machines and add tabs into vanilla recipe book

- custom recipe book category can be registered via forge registries.
- constructor have `IRecipeType<?>` to set where recipe book category belongs
- modder can set if category is `search` - no filtering is applied for displayed recipes
- `setPredicate(IRecipe<?>)` non-search category with filtering
- `setIcon(ItemStack...)` recipe book category icon
```
    @SubscribeEvent
    public static void registerRecipeBookCategories(RegistryEvent.Register<RecipeBookCategories> event) {
        event.getRegistry().register(new RecipeBookCategories(TestRecipe.millstone).setSearch().setIcon(new ItemStack(Items.BELL)).setRegistryName("base_crafting_book_test", "test"));
    }
```
- Custom machine container must extends `RecipeBookContainer<C extends IInventory>` 
- Custom machine gui just render and handle recipe book as in e.g. AbstractFurnaceScreen

### CODE REVIEW

File: patches/minecraft/net/minecraft/client/util/RecipeBookCategories.java.patch
Comment (TheSilkMiner) [inline patches/minecraft/net/minecraft/client/util/RecipeBookCategories.java.patch]: This breaks binary compatibility

File: patches/minecraft/net/minecraft/client/gui/recipebook/RecipeBookGui.java.patch
Comment (TheSilkMiner) [inline patches/minecraft/net/minecraft/client/gui/recipebook/RecipeBookGui.java.patch]: Don't use `import`s in patches

File: patches/minecraft/net/minecraft/client/gui/recipebook/RecipeBookGui.java.patch
Comment (TheSilkMiner) [inline patches/minecraft/net/minecraft/client/gui/recipebook/RecipeBookGui.java.patch]: You should move this to another file IMO.
Patches are supposed to be minimal and adding classes is kinda against that

File: patches/minecraft/net/minecraft/inventory/container/AbstractFurnaceContainer.java.patch
Comment (TheSilkMiner) [inline patches/minecraft/net/minecraft/inventory/container/AbstractFurnaceContainer.java.patch]: You should move this over to the class where you added this method and use `instanceof` checks. It's not the best OOP design, but it saves up on patch count. Same for `PlayerContainer`, `WorkbenchContainer`

File: patches/minecraft/net/minecraft/inventory/container/RecipeBookContainer.java.patch
Comment (TheSilkMiner) [inline patches/minecraft/net/minecraft/inventory/container/RecipeBookContainer.java.patch]: Don't use imports in patches

File: patches/minecraft/net/minecraft/client/util/RecipeBookCategories.java.patch
Comment (yanny7) [reply patches/minecraft/net/minecraft/client/util/RecipeBookCategories.java.patch]: Any idea how to do it without breaking compatibility?

File: patches/minecraft/net/minecraft/client/gui/recipebook/RecipeBookGui.java.patch
Comment (yanny7) [reply patches/minecraft/net/minecraft/client/gui/recipebook/RecipeBookGui.java.patch]: removed imports from patches

File: patches/minecraft/net/minecraft/client/gui/recipebook/RecipeBookGui.java.patch
Comment (yanny7) [reply patches/minecraft/net/minecraft/client/gui/recipebook/RecipeBookGui.java.patch]: Moved UpDownButton to new class

File: patches/minecraft/net/minecraft/inventory/container/AbstractFurnaceContainer.java.patch
Comment (yanny7) [reply patches/minecraft/net/minecraft/inventory/container/AbstractFurnaceContainer.java.patch]: implemented default body for getRecipeType

File: patches/minecraft/net/minecraft/inventory/container/RecipeBookContainer.java.patch
Comment (yanny7) [reply patches/minecraft/net/minecraft/inventory/container/RecipeBookContainer.java.patch]: Fixed

File: patches/minecraft/net/minecraft/client/util/RecipeBookCategories.java.patch
Comment (yanny7) [reply patches/minecraft/net/minecraft/client/util/RecipeBookCategories.java.patch]: Fixed ABI

**Review by TheSilkMiner — CHANGES_REQUESTED** [review]
There are still some breaking changes within this PR:
- `RecipeTabToggleWidget#field_193921_u`'s type has been changed, along with the constructor references
- `RecipeTabToggleWidget#func_201503_d`'s return type has been changed
- `ClientRecipeBook#field_197931_e` had its type changed
- `ClientRecipeBook#func_216767_a` had its parameters changed
- `ClientRecipeBook#func_202891_a` was deleted and replaced with another method

Also, I am not a fan of replacing constructor arguments inside an `enum`, since that leads to a much bigger set of changes in patches.

The way I'd approach this is rather have the recipe tab toggle widget category remain the same, but redirect all usages towards another `IForgeRecipeBookCategory<?>` , which gets assigned by vanilla in the constructor. Also, provide a secondary constructor, and patch `RecipeBookGui#initSearchBar` to use the new one. which would set `field_193921_u` to `null` and actually reference the category.
Meanwhile, `func_201503_d` can be deprecated and replaced with an additional function, patching the call when needed. 

Regarding `ClientRecipeBook`, I'd do the same: deprecate the current methods, try to patch vanilla to use the newly added ones, and have the old method redirect to the new method if possible, so that patch size can be lowered if needed (E.g. `func_216767_a` could be redirected to a `getRecipeByCategory` method with the same body as much as possible).

Regarding the `enum`, I'd hardcode the types directly within the constructor: this way there is no need to modify the constructor. This shouldn't matter much, but it may lead to lower patch sizes. Or probably just hardcoding the return type in the method, by making `IForgeRecipeBookCategory` provide default implementations for most of the methods you added (`isSearch`, `getPredicate`, `getRecipeType` and friends).

Nevertheless, these suggestions may need some further discussion around feasibility, since if they end up complicating the code too much, maybe it's not that worth it.

**Comment by ichttt** [comment]
Just as a note: https://github.com/MinecraftForge/MinecraftForge/pull/6669 exists, but it seems abandond.
Also, new features should first be PRd to 1.15 instead of 1.14

**Comment by yanny7** [comment]
I port this PR to 1.15 after will be approved for 1.14

**Comment by diesieben07** [comment]
> I port this PR to 1.15 after will be approved for 1.14

That is the wrong direction. New features go in 1.15 first. 1.14 is LTS and receives backports after the fact.

**Comment by yanny7** [comment]
Its enough for me that there will be no problem with code for 1.14. I dont want to waste my time by porting it right now into 1.15 when code is not polished. I have no problem that after approving that code is OK for merging this PR will be postproned until 1.15 port will be available

**Comment by diesieben07** [comment]
A feature PR for 1.14 won't get approved if it's not a backport of a 1.15 feature.

**Comment by yanny7** [comment]
Feature was ported to 1.15

**Comment by yanny7** [comment]
Still WIP

**Comment by TheCurle** [comment]
This 1.14 PR has a 1.15 version and is unnecessary. 
As a side note, the same applies as before. I don't see a 1.16 version, so you should target that before the 1.15 PR is considered.
Apologies for the hassle, but manpower is limited here...

**Comment by CREEATION** [comment]
@yanny7 any news on this? 🐢

**Pull request closed without merging**
